### PR TITLE
Train textcat with config

### PR DESCRIPTION
**Pull request by svlandeg**
[Update 18 March: requires [Thinc 8.0.0a3](https://github.com/explosion/thinc/pull/327)]

## Description
Adopted `textcat` for the new 3.0 design:
* Found & fixed a few bugs in the new implementation. Requires `thinc` > 8.0.0a2 with the new `softmax_activation`
* Implemented the old `build_text_classifier` and use that as current default
* added `textcat_bow_defaults.cfg` and `textcat_cnn_defaults.cfg` to reflect the default implementations on `master` for those specific architectures
* added textcat unit tests for a variety of possible configs (which proved to be extremely useful during development)

Changes to `train_textcat.py:`
* build the NLP model from a config file + provide a default to make the transition to config-based as easy as possible
* support for multi-class and multi-label datasets
* enable specifying a different dataset that gets loaded with `ml_datasets.loaders()` by using `-d`
* enable specifying a threshold for multi-label classification. Labels with an occurrence count below that threshold are discarded from training.

With respect to Tok2Vec:
* revert shenanigans with `_set_dims` for Tok2Vec as initialization can now be done with `tok2vec.initialize(X=docs)`

## Evaluation:
To ensure there are no `textcat` regressions moving from Thinc 7 / spaCy 2 to Thinc 8 / spaCy 3, I performed several tests. Showing in the table the best F-score across 50 epochs, using 2000 training articles, e.g. `train_textcat.py train_textcat_config.cfg -n 50 -d dbpedia`.

Datasets:
 * IMDB: **binary classification**
 * DBpedia: **multi-class, single-label classification**

| Dataset | Architecture | spaCy 2 (current `master`) | spaCy 3 (current PR) |
| ------- | ------------ | -------------------------- | -------------------- |
| IMDB    | BOW          | 78.5 %                       | 81.4 %                 |
| IMDB    | CNN          | 79.1 %                       | 79.4 %                 |
| IMDB    | DEFAULT      | 76.4 %                       | 77.1 %                 |
| DBpedia | BOW          | 90.5 %                       | 90.7 %                 |
| DBpedia | CNN          | 84.8 %                       | 86.4 %                 |
| DBpedia | DEFAULT      | 89.7 %                       | 90.0 %                 |   

This clearly shows there are no regressions - on the contrary, the new implementations seem to work slightly better, even though we attempted to make a 1-1 mapping. I assume Thinc 8 has some issues fixed from the previous version.

I've also done some preliminary experiments on a new **multi-class, multi-label** dataset, CMU (cf https://github.com/explosion/ml-datasets/pull/1). It has >300 labels but a threshold of 20 was used to keep only the 66 most frequent labels.

| Dataset | Architecture  | spaCy 3 (current PR) |
| ------- | ------------ | -------------------- | 
| CMU | BOW                       | 21.7 %                 |
| CMU | CNN                       | 30.5 %                 |
| CMU | DEFAULT                 | 14.4 %                |   

The CNN result of 30.5% seems to be in line with the first result (31.5%) reported in [this blog](https://www.analyticsvidhya.com/blog/2019/04/predicting-movie-genres-nlp-multi-label-classification/), where they further boosted performance to 43.7% by playing with the threshold for determing a "true" prediction. So it looks encouraging, but definitely some more work could be done on this.

### Types of change
enhancement

## Checklist
- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by svlandeg** [comment]
We'll need to release Thinc 8.0.0a2 first before the tests can run...

File: spacy/ml/models/textcat.py [hunk 1]
```diff
@@ -33,13 +36,99 @@ def build_simple_cnn_text_classifier(tok2vec, exclusive_classes, nO=None):
...
+                >> list2ragged()
+                >> ParametricAttention(width)
+                >> reduce_sum()
+                >> residual(Maxout(nO=width, nI=width))
+                >> Linear(nO=nO, nI=width)
+                >> Dropout(0.0)
+        )
+
+        linear_model = build_bow_text_classifier(
+            nO=nO, ngram_size=ngram_size, exclusive_classes=False, no_output_layer=False
```
Comment (adrianeboyd) [inline spacy/ml/models/textcat.py]: Should this be `exclusive_classes=exclusive_classes`?
